brainframe/cli/config.py

```diff
@@ -0,0 +1,73 @@
+import os
+from distutils.util import strtobool
+from pathlib import Path
+from typing import Callable, Dict, Generic, Optional, TypeVar
+
+import yaml
+
+from . import print_utils
+
+_DEFAULTS_FILE_PATH = Path(__file__).parent / "defaults.yaml"
+
+
+T = TypeVar("T")
+
+
+class Option(Generic[T]):
+    """A configuration option.
+
+    Option values are determined using the following precedence:
+      1. From an environment variable with the name "BRAINFRAME_" followed by the option
+         name in all caps
+      2. From the defaults file shipped with this distribution
+      3. The fallback value, which is None
+    """
+
+    name: str
+    value: Optional[T] = None
+    default: Optional[T] = None
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def load(
+        self, converter: Callable[[str], T], defaults: Dict[str, str]
+    ) -> None:
+        default = defaults.get(self.name)
+        env_var_name = "BRAINFRAME_" + self.name.upper()
+
+        value: Optional[str]
+        if env_var_name in os.environ:
+            value = os.environ[env_var_name]
+        else:
+            value = default
+
+        self.value = None if value is None else converter(value)
+        self.default = None if default is None else converter(default)
+
+
+install_path = Option[Path]("install_path")
+data_path = Option[Path]("data_path")
+
+is_staging = Option[bool]("staging")
+staging_username = Option[str]("staging_username")
+staging_password = Option[str]("staging_password")
+
+
+def load() -> None:
+    """Initializes configuration options"""
+    if not _DEFAULTS_FILE_PATH.is_file():
+        print_utils.fail_translate(
+            "general.missing-defaults-file",
+            defaults_file_path=_DEFAULTS_FILE_PATH,
+        )
+
+    with _DEFAULTS_FILE_PATH.open("r") as defaults_file:
+        defaults = yaml.load(defaults_file, Loader=yaml.FullLoader)
+
+    install_path.load(Path, defaults)
```

Comment:
It seems mildly redundant to mention the type twice (in the option definition and again in load time). 

Would it make more sense to pass the type in, similar to how parser.add_argument(type=callable, ...) typically works?

Reply:
I wasn't super happy about this either, but unfortunately the type of the option is not always the same as the callable needed to convert it to that type. For example, with `bool`, we need to use `distutils.util.strtobool` because converting a string with `bool()` doesn't have the semantics we want.

Reply:
Doesn't YAML have first-class support for bools though? :thinking:

Reply:
Yes, but since these configuration options can also be set via environment variable, we can't rely on receiving a properly typed bool as a value.